### ISSUE TITLE
Performance - getDiskStores forces me to discard an array every tick!

### ISSUE THREAD

**Issue by cjbrowne** [oshi/oshi#499]
Hi there, me again!  I thought I'd make an issue this time in case I've missed something obvious again so I don't put in unnecessary work here.

The issue is that in my application I am aiming for bare minimal CPU and memory usage, which means I'm stressing OSHI a bit here and there.

This particular case deals with `HardwareAbstractionLayer.getDiskStores()`.  I have code something like the followig:

```
HardwareAbstractionLayer hardware = systemInfo.getHardware();
HWDiskStore[] diskStores = hardware.getDiskStores();
while(true) {
  diskStores = hardware.getDiskStores();

  // [skipped] copy each HWDiskStore's vital statistics into a DTO

  Thread.sleep(100L); // there's a try block around this but it's irrelevant
}
```

This code causes a large amount of old `HWDiskStore[]` (~10 per second) to build up until the GC runs and inevitably decides not to bother cleaning up most of them (for no reason that I can understand, but GC's gonna GC).  This in turn causes longer GC pauses.

It also seems to be quite CPU intensive in and of itself, because it needs to re-construct the `HWDiskStore[]` on every loop, rather than simply updating the existing `HWDiskStore[]` that I already have.  Even though there's only one `HWDiskStore` in this system, it's causing a noticeable CPU usage spike every time the loop runs.

When it comes to network interfaces, there is a `NetworkIF.updateNetworkStats` call that allows me to skip re-creating the entire `NetworkIF[]` on every loop - does it make any sense to implement something similar for `HWDiskStore`, or is there a different suggestion you might make for my case where I can avoid re-creating the array?

For now I have tuned the poll rate down from 10x per second to 1, which is sufficient to make the performance tolerable without negatively impacting the functionality, but it would be nice to be able to poll the disk usage more often if I can.  The use-case is monitoring for a very high throughput service which should be unaffected by the monitoring but the monitoring should also be able to trigger alerts quickly if a problem happens and have a (fairly) accurate timestamp for when the problem started so that it can be cross-referenced against what part of the pipeline was in use at the time (this is where increasing the poll rate comes in - the faster I poll, the more accurate the timestamp can be, and while I will never hope to get sub-millisecond precision here I would at least like to get sub-second precision).

I do intend to eventually replace this with something written in a faster language like Go or C++, but OSHI is proving immensely valuable in coding the Java POC.

**Comment by dbwiddis** [comment]
Yes, we know this -- #310 discusses the widespread problem.   #365 discusses a long term means of fixing it.

However, I'm not sure the redesign is happening any time soon and it seems reasonable to try to fix this in the current API.   Also, similar to the recent OSProcess rewrite (#489), a lot of the latency with disks is in WMI and it's possible to cache unchanging info and only fetch more recent info, even without an explicit "Update" method.

I'd welcome any attempt to improve this.

**Comment by YoshiEnVerde** [comment]
I've been swamped by work the last few months.
I haven't been able to even peek at #365 since late last year.

Go ahead and update anything as needed.
Whenever #365's MVP is ready, all these fixes should be easy to add to it (since that was the idea behind that code refactoring)

**Comment by dbwiddis** [comment]
Let's go with a band-aid fix of adding an update method to the `HWDiskStore` similar to the `NetworkIF` class.  You won't detect new drives this way (but you can periodically poll for that) and you'll have to protect against a drive disappearing on you.

**Comment by dbwiddis** [comment]
Fixed in #501
